cmd/juju/model/destroy.go

```diff
@@ -258,13 +258,21 @@ func (c *destroyCommand) Run(ctx *cmd.Context) error {
 	}
 	defer func() { _ = api.Close() }()
 
+	modelStatuses, err := api.ModelStatus(names.NewModelTag(modelDetails.ModelUUID))
+	if err != nil {
+		return errors.Annotate(err, "getting model status")
+	}
+	modelStatus := modelStatuses[0]
+
+	// Error out if the model has detachable storage and is not authorized to destroy or release it.
+	detachableVolumes, detachableFilesystems := countDetachableStorage(modelStatus)
+	if (detachableVolumes > 0 || detachableFilesystems > 0) && !(c.destroyStorage || c.releaseStorage) {
+		return handlePersistentStorageError(names.NewModelTag(modelDetails.ModelUUID), modelName, api)
```

Comment:
Some more refactoring is needed here. The handlePersistentStorageError method is recalulating the detachable storage counts and worse, taking an api so that it can again call model status. We've already got the info. I'd pass that into the "handlePersistentStorageError" and do the counting etc in there.

Reply:
Agreed.  It felt a bit awkward simply passing that info into `handlePersistentStorageError` because that function also gets called by `handleError`, which hasn't queried the model yet.  Instead I encapsulated `getModelStatus` and `generatePersistentStorageErrorMsg` so they could be used in both places and we avoid the double api call.  wdyt?